packages/flutter_tools/lib/src/runner/target_devices.dart

```diff
@@ -345,11 +345,19 @@ class TargetDevices {
 
   Future<Device> _chooseOneOfAvailableDevices(List<Device> devices) async {
     _displayDeviceOptions(devices);
-    final String userInput =  await _readUserInput(devices.length);
-    if (userInput.toLowerCase() == 'q') {
-      throwToolExit('');
+
+    Device? chosenDevice;
+    while (chosenDevice == null) {
+      final String userInputString = await _readUserInput(devices.length);
+      if (userInputString.toLowerCase() == 'q') {
+        throwToolExit('');
+      }
+      final int deviceIndex = int.parse(userInputString) - 1;
+      if (deviceIndex > -1 && deviceIndex < devices.length) {
+        chosenDevice = devices[deviceIndex];
+      }
     }
-    return devices[int.parse(userInput) - 1];
+    return chosenDevice;
```

Comment:
Yeah, although they're similar, they're separate because `TargetDevices` knows in advance exactly what numbers are valid options, whereas `TargetDevicesWithExtendedWirelessDeviceDiscovery` does not since options get added asynchronously. 

If you look into the `TargetDevices._readUserInput` function, it uses `terminal.promptForCharInput`, which handles the while loop and validates it's an acceptable input, so you shouldn't need to change this.

Since `TargetDevicesWithExtendedWirelessDeviceDiscovery` does not know in advance what options are valid, it can't use `terminal.promptForCharInput`, so it accepts any number as input and then validates it's a valid option.

Reply:
Ah good call, I will revert the changes made to `TargetDevices` then and find a test that uses `TargetDevicesWithExtendedWirelessDeviceDiscovery`. I think the test I tried to piggy back on was using `TargetDevices` method for selecting a device